PRINCIPLES.md

```diff
@@ -0,0 +1,254 @@
+
+Secondly, verifying the correctness and self-consistency of data is significantly less complex than verifying the correctness of a program's behaviour, which is fundamentally undecidable. 
+Verifying that a set of declarations is correct, however, _is_ decidable, even if sometimes computationally expensive.
+As a general rule, this implies that the data language used to defined the Desired State should have no control flow structure, and consist exclusively of referentially transparent expressions without side effects. 
+In other words, a human-readable data-serialization format.
+
+It is also preferrable to use a widely supported language to define the Desired State for the sake of interoperability. For example, YAML, XML, JSON and TOML all have broad support and well defined specifications, although there are many other suitable candidates.
+
+#### Why is human readability required?
+
+Operationally, it has been proven time and time again that the canonical Desired State of a system should be human-readable and writable.
+
+The Desired State encodes the _intent_ of a system. For example, where internet traffic should be routed. The collection of decisions about a system captured in its Desired State are of particular interest to its human operators. They must be able to directly describe their intent about the state of a system; and similarly, must be able to decipher the intended state of a system from its configuration.
+
+This principles not only capture a requirement about the format of the configuration, but also the qualitative readability of configuration _in practice_. For example, a simple YAML or XML file can be easily read, understood and modified by human operators, but experience has taught us that these formats also allow the creation of complex self-referential documents beyond the ability of most humans to interpret. Such complex document would violate this principles, even though the formats are, on the surface, human-readable.
+
+This relative readability also implies that the users are relevant when evaluating whether a system follows this principle. For example, a Desired State defined with a rich grammar of S-Expressions would suite a team of developers with a background in functional programming, but would violate this principle if the majority of the team found the format incomprehensible.
+
+Having a human-readable Desired State does not in any way preclude the use of rich tooling or graphical interfaces that facilitate Desired State generation and interpretation. It only require that the canonical source of truth be human readable and writable plain text.
+
+#### How much of a system must be declared?
+
+Ideally, all of it; and the entire system can be recreated exclusively from its Desired State.
```

Comment:
I think this is confusing in the context of application data. If I really wanted to recreate an entire system I'd have to consider recreating the application data that it requires to operate, and may even be interpreted in the context of disaster recovery (is GitOps a way to recover the latest known state of a system should disaster strike?) This also moves the conversation into scenarios that are perhaps beyond configuration, for instance, should DB migrations be declared as part of the desired state, and should I want to recreate a system matching that state must GitOps guarantee db schema consistency?

Reply:
Definitely should mention that DB Migrations Question in a FAQ on the Website. Will be a commonly asked question.

Reply:
> is GitOps a way to recover the latest known state of a system should disaster strike?

The _latest configured behaviour_? Yes. The latest state including application data? Not at this time.

> Definitely should mention that DB Migrations Question in a FAQ on the Website. Will be a commonly asked question.

:+1: That's a super common discussion.

Reply:
> system matching that state must GitOps guarantee db schema consistency?

I think so, although this is genuinely difficult currently. Or rather, the db schema at a point in time should be declared. Moving to that schema from some arbitrary starting point is detail left to the reconciling agent. Currently, this will be in the form of a set of unidirectional migrations generated manually. That's an implementation detail though.

Reply:
I agree that we should mention application data here though.

Reply:
@murillodigital wrote:
> should DB migrations be declared as part of the desired state?

Yes.

-----

@murillodigital wrote:
> must GitOps guarantee db schema consistency?

Should instead of must.


GitOps should guarantee database schema consistency.

-----

@bricef wrote:
> I think so, although this is genuinely difficult currently. Or rather, the db schema at a point in time should be declared. Moving to that schema from some arbitrary starting point is detail left to the reconciling agent. Currently, this will be in the form of a set of unidirectional migrations generated manually. That's an implementation detail though.

Agreed that database schema at a point in time should be declared.

Database migrations should be generated automatically, instead of manually.

Examples of implementation details:
1. [Ask HN: How does your development team handle database migrations?](https://news.ycombinator.com/item?id=19880334)
2. [Evolutionary Database Design](https://www.martinfowler.com/articles/evodb.html) by @pramodsadalage and @martinfowler at @thoughtworks using tools such as:
• https://github.com/liquibase/liquibase by @liquibase and @Datical
• https://github.com/flyway/flyway by @flyway and @red-gate
• https://github.com/mybatis/migrations by @mybatis
3. https://github.com/schemahero/schemahero by @schemahero, [donated by](https://www.replicated.com/blog/replicated-donates-schemahero-to-the-cncf/) @replicatedhq to @cncf
-----

From [GitOps Principles Committee - April 14, 2021 after 40:50](https://www.youtube.com/watch?v=zWxAZuHNGYM&t=2450s)

@bricef said:
> It's more about an aspirational direction, not necessarily what we do right now.

Agreed.

-----

@tonit wrote:
> DB Migrations Question in a FAQ

Agreed.